src/hotspot/share/tsan/tsanOopMapTable.hpp

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2024, Google and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_VM_PRIMS_TSAN_OOPMAPTABLE_HPP
+#define SHARE_VM_PRIMS_TSAN_OOPMAPTABLE_HPP
+
+#include "gc/shared/collectedHeap.hpp"
+#include "memory/allocation.hpp"
+#include "oops/oop.inline.hpp"
+#include "oops/weakHandle.hpp"
+#include "tsan/tsanOopMap.hpp"
+#include "utilities/resizeableResourceHash.hpp"
+
+namespace TsanOopMapImpl {
+
+  struct PendingMove {
+    char *source_begin() const { return source_address; }
+    char *source_end() const { return source_address + n_bytes; }
+    char *target_begin() const { return target_address; }
+    char *target_end() const { return target_address + n_bytes; }
+    char *source_address;
+    char *target_address;
+    size_t n_bytes;  // number of bytes being moved
+  };
+
+}  // namespace TsanOopMapImpl
+
+// For tracking the lifecycle (alloc/move/free) of interesting oops
+// that tsan needs to know.
+class TsanOopMapTableKey : public CHeapObj<mtInternal> {
+ private:
+  WeakHandle _wh;
+
+  // Address of the oop tracked by the WeakHandle.
+  // After an object is freed, the WeakHandle points to null oop. We
+  // need to cache the original oop address for notifying Tsan after
+  // the object is freed.
+  oopDesc *_obj;
+
+ public:
+  TsanOopMapTableKey(oop obj);
+  TsanOopMapTableKey(const TsanOopMapTableKey& src);
+  TsanOopMapTableKey& operator=(const TsanOopMapTableKey&) = delete;
+
+  void release_weak_handle() const;
+  oop object_no_keepalive() const;
```

Comment:
Alternatively, it might be OK to declare TsanOopMapTableKey::_obj as an `oop` instead of `oopDesc*`. The main point is to avoid type conversions in various places.

Reply:
Using `oop` for `TsanOopMapTableKey::_obj` is ok.

I also just built and tested with `fastdebug` JDK binary. I found a few places need `cast_from_oop<type>` conversions, e.g. in `InstanceKlass::initialize` when calling `SharedRuntime::tsan_acquire`. I've fixed those. I'll push my changes shortly. 

The tsan jtreg tests pass with `fastdebug` binary, except `RawRacyNativeLoopTest.java` and `NonRacyCasLoopTest.java`. That's the same with `slowdebug` binary. Those two tests fail with following assert. The failures should be unrelated to the tsan oop map. There is a separate internal bug that I created earlier to this specific issue.

```
#  Internal Error (/usr/local/google/home/jianglizhou/github/tsan-oopMap-weakhandle/src/hotspot/share/oops/accessBackend.cpp:224), pid=1986120, tid=1986364
#  assert(state == _thread_in_vm || state == _thread_in_Java || state == _thread_new) failed: Wrong thread state for accesses: 4
```